sdk/core/azure-core-http-httpurlconnection/src/main/java/com/azure/core/http/httpurlconnection/SocketClient.java

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * A socket client used for making PATCH requests
+ */
+
+class SocketClient {
+
+    private static final String HTTP_VERSION = " HTTP/1.1";
+
+    /**
+     * Opens a socket connection, then writes the PATCH request across the
+     * connection and reads the response
+     *
+     * @param httpRequest The HTTP Request being sent
+     * @return an instance of HttpUrlConnectionResponse
+     */
+    public static HttpUrlConnectionResponse sendPatchRequest(HttpRequest httpRequest) throws IOException {
+        String host = httpRequest.getUrl().getHost();
+        int port = httpRequest.getUrl().getPort();
+        if (httpRequest.getUrl().getProtocol().equals("https")) {
+            SSLSocketFactory sslSocketFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
+            try (SSLSocket socket = (SSLSocket) sslSocketFactory.createSocket(host, port)) {
+                return doInputOutput(httpRequest, socket);
+            }
+        } else if (httpRequest.getUrl().getProtocol().equals("http")) {
+            try (Socket socket = new Socket(host, port)) {
+                return doInputOutput(httpRequest, socket);
+            }
+        }
+        throw new ProtocolException("Only HTTP and HTTPS are supported by this client.");
+    }
+
+    /**
+     * Calls buildAndSend to send a String representation of the request across the output
+     * stream, then calls buildResponse to get an instance of HttpUrlConnectionResponse
+     * from the input stream
+     *
+     * @param httpRequest The HTTP Request being sent
+     * @param socket An instance of the SocketClient
+     * @return an instance of HttpUrlConnectionResponse
+     */
+    private static HttpUrlConnectionResponse doInputOutput(HttpRequest httpRequest, Socket socket) throws IOException {
+        httpRequest.setHeader(HttpHeaderName.HOST, httpRequest.getUrl().getHost());
+        if (!"keep-alive".equalsIgnoreCase(httpRequest.getHeaders().getValue(HttpHeaderName.CONNECTION))) {
+            httpRequest.setHeader(HttpHeaderName.CONNECTION, "close");
+        }
+
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
+             OutputStreamWriter out = new OutputStreamWriter(socket.getOutputStream())) {
+
+            buildAndSend(httpRequest, out);
+            HttpUrlConnectionResponse response = buildResponse(httpRequest, in);
+
+            String redirectLocation = response.getHeaders().stream()
+                .filter(h -> h.getName().equals("Location"))
+                .map(HttpHeader::getValue)
+                .findFirst()
+                .orElse(null);
+
+            if (redirectLocation != null) {
+                if (!redirectLocation.startsWith("http")) {
+                    httpRequest.setUrl(new URL(httpRequest.getUrl(), redirectLocation));
+                } else {
+                    httpRequest.setUrl(redirectLocation);
+                }
```

Comment:
Just in terms of readability, it is better to avoid unnecessary boolean negations. Here, instead of testing "if redirect location does NOT start with http", you could test with "if redirect location does start with http", and just flip the lines 84 and 86.